py_gtfs_rt_ingestion/dev_test_setup.py

```diff
@@ -0,0 +1,298 @@
+
+    return SetupArgs(**vars(parser.parse_args(args)))
+
+
+def del_objs(del_list: list[str], bucket: str) -> int:
+    """
+    Delete del_list of objects from bucket
+    """
+    s3_client = boto3.client("s3")
+    delete_objs = {
+        "Objects": [
+            {"Key": uri.replace(f"s3://{bucket}/", "")} for uri in del_list
+        ]
+    }
+    try:
+        s3_client.delete_objects(
+            Bucket=bucket,
+            Delete=delete_objs,
+        )
+    except Exception as e:
+        logging.exception(e)
+        return 0
+    return len(del_list)
+
+
+def make_del_jobs(files: list[str], bucket: str) -> int:
+    """
+    Spin up Multiprocessing jobs to delete batches of objects
+    """
+    chunk = 750
+    with Pool(os.cpu_count()) as pool:
+        l_t = [
+            (files[i : i + chunk], bucket) for i in range(0, len(files), chunk)
+        ]
+        results = pool.starmap_async(del_objs, l_t)
+        del_count = sum(r for r in results.get())
+    return del_count
+
+
+def get_del_obj_list(bucket: str, uri_root: str) -> list[str]:
+    """
+    Return list of objects in bucket, removing uri_root object from return list
+    """
+    print(f"Checking for objects to delete in {uri_root}...")
+    files_to_delete = []
+    for (uri, _) in file_list_from_s3(bucket, LAMP_PREFIX):
+        files_to_delete.append(uri)
+    # Remove uri_root from list so that root directory is not deleted.
+    files_to_delete.remove(uri_root)
+    return sorted(files_to_delete, reverse=True)
+
+
+def clear_dev_buckets(args: SetupArgs) -> None:
+    """
+    Clear all development buckets of objects
+    """
+    bucket_list = (
+        DEV_INGEST_BUCKET,
+        DEV_EXPORT_BUCKET,
+        DEV_ARCHIVE_BUCKET,
+        DEV_ERROR_BUCKET,
+    )
+
+    for bucket in bucket_list:
+        action = None
+        uri_root = f"s3://{os.path.join(bucket, LAMP_PREFIX)}"
+        files_to_delete = get_del_obj_list(bucket, uri_root)
+        if len(files_to_delete) == 0:
+            print("No objects found... skipping bucket.")
+        else:
+            while action not in ("n", "no"):
+                # Print list of bucket objects
+                if action in ("list", "ls"):
+                    print(f"/{'*' * 50}/")
+                    for uri in files_to_delete:
+                        print(f"{uri.replace(uri_root, '')}")
+                    print(f"/{'*' * 50}/")
+                # Proceed with object deletion
+                elif action in ("y", "yes") or args.force_delete:
+                    delete_count = make_del_jobs(files_to_delete, bucket)
+                    # If not all objects deleted, retry
+                    if delete_count < len(files_to_delete):
+                        print(
+                            f"Only {delete_count:,} of {len(files_to_delete):,}"
+                            " deleted... will retry."
+                        )
+                        files_to_delete = get_del_obj_list(bucket, uri_root)
```

Comment:
was this a case that you were regularly seeing?

also, `logging` again rather than print.

Reply:
It is intermittent depending on the number of objects that are being deleted and also if boto3 just randomly craps out because of a connection issue. For less than 5,000 objects it almost never happens but for 20,000+ it seems to happen about 20% of the time.